src/web/components/SlotBodyEnd.tsx

```diff
@@ -38,6 +39,18 @@ export const SlotBodyEnd = ({
     // Load the view log from localStorage so it can be added to the Contributions payload
     const epicViewLog = getViewLog();
 
+    // These are expensive - only retrieve them once
+    const mvtId = useMemo(() => Number(getCookie('GU_mvt_id')), []);
+    const lastOneOffContributionDate = useMemo(
+        getLastOneOffContributionDate,
+        [],
+    );
+    const showSupportMessaging = useMemo(shouldShowSupportMessaging, []);
+    const recurringContributor = useMemo(
+        () => isRecurringContributor(isSignedIn),
+        [],
```

Comment:
shouldnt this change when `isSignedIn` changes?

Reply:
That's a good question, I'd also assume that it would.

Reply:
cc @oliverlloyd too - does this get rendered server side initially? I.e. is there a chance we'll memoise the cookie values before they are accessible?

We only ever want to render the slot once (even if data subsequently changes because of a user action).

Reply:
> does this get rendered server side initially? ...

it appears to be only rendered in the browser

Reply:
So basically, what William said. Because the component is rendered as a 'Portal' it only appears on the client.

Plus, it's worth pointing out that this component will rerender several times:

Once: with signedIn = `undefined` and countryCode = `undefined`
Twice with countryCode = 'UK' say
and (if they are logged in) Thrice with signedIn 'true'

Is 'thrice' a word?

Reply:
And, you should add `isSignedIn` to the deps array to be sematically correct for useMemo

Reply:
Interesting Twitter convo: https://twitter.com/ryanflorence/status/1113084393134809090

Reply:
I just chatted with Oliver IRL. Like William says, this will only ever be rendered in the browser (it's within a Portal) but it will be re-rendered up to 3 times:

1. Initial render
2. When `countryCode` gets set
3. When `isSignedIn` changes (if user is signed in)

It sounds like this isn't quite what we want.

Reply:
Okay thanks all. How do we avoid calling Slot Machine more than once than? And only with the 'right' data? It seems non-obvious given that the component doesn't know whether country code or signed in status will change. 

cc @oliverlloyd @tjmw @liywjl

Reply:
If you mean preventing renders, you can't. React wants you to build things such that rerenders can happen without it being a problem.

In terms of preventing the cookie read (if that's what you meant, which I suspect is the case), then you can pass an empty array as the deps - this 'guarantees' it will only fire once but if the call does actually have dependencies then technically this is invalid.

This problem space is explained really well here: https://overreacted.io/a-complete-guide-to-useeffect/#dont-lie-to-react-about-dependencies

If we really want to keep the deps array true, as per ^, then the question is why not make the read/call twice? It's true that there's a performance hit but if the user has signed in or out we should update the page contents to reflect this, no?

Reply:
I'm not sure how valid this is (hopefully it'll come out in the PR view, once I've opened one) but in this wip branch here I've got an effect with dependencies which I only want to happen once, and I'm keeping track of it with an explicit piece of state (`hasSentNotification`) which I set within the hook. It seems to work well. Could that be an option?

Reply:
@oliverlloyd I mean really in terms of the call to the Slot Machine service.

Reply:
Ah, so you want to delay the api call until `isVisible` (and maybe some other things) and then only make it once?

In _that_ case, I think you should compose a separate component and have that as a child. The parent can be the one that rerenders accepting the state changes, and only when the state is how you want it (`isVisible`) then it renders the child. 

Inside the child you can have a useEffect with a valid empty deps array that means it will only run once.

If that's not clear I'm happy to knock up a quick example